modules/gapi/src/compiler/gcompiler.cpp

```diff
@@ -317,44 +414,20 @@ void cv::gimpl::GCompiler::validateInputMeta()
                      "got " + std::to_string(m_metas.size()) + " meta arguments)"));
     }
 
-    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto) {
-        switch (proto.index())
-        {
-        // FIXME: Auto-generate methods like this from traits:
-        case GProtoArg::index_of<cv::GMat>():
-        case GProtoArg::index_of<cv::GMatP>():
-            return util::holds_alternative<cv::GMatDesc>(meta);
-
-        case GProtoArg::index_of<cv::GFrame>():
-            return util::holds_alternative<cv::GFrameDesc>(meta);
-
-        case GProtoArg::index_of<cv::GScalar>():
-            return util::holds_alternative<cv::GScalarDesc>(meta);
-
-        case GProtoArg::index_of<cv::detail::GArrayU>():
-            return util::holds_alternative<cv::GArrayDesc>(meta);
-
-        case GProtoArg::index_of<cv::detail::GOpaqueU>():
-            return util::holds_alternative<cv::GOpaqueDesc>(meta);
-
-        default:
-            GAPI_Assert(false);
-        }
-        return false; // should never happen
-    };
```

Comment:
Amount of code removed here and added atop of that gives some hint... :)

Reply:
That's the controversial point of having a visitor as some class/etc. -> while previously we could use plain switch (well, not very automated but integrated perfectly into its environment), with a distinct _class_ the handling logic is torn away from the context - we don't have scope closures for such objects like we could have with lambdas, for example.

Would it be possible to construct an ad-hoc visitor with typed lambdas right in place? It would be a game changer then.

I believe the limitation I've got into was that the `visit` return type was determined by the visitor lambda called - and it could vary; but I believe we could relax such requirement here and now.

Reply:
>That's the controversial point of having a visitor as some class/etc. -> while previously we could use **plain switch**

in current piece of code we need two stacked `switch`s  similar with

```
  switch(expected)
  {
     case A:
            ...
            switch (got)
            {
                case AA:
                     PRINT "A and AA mismatch"
                case BB:
                     PRINT "A and BB mismatch"
            }
     case B:
        
          switch (got)
              case AA:
                      PRINT "B and AA mismatch"
  }
```
In current implementation it solved by using two simple classes


>>I believe the limitation I've got into was that the visit return type was determined

exactly! it is the problem i faced ( without writting  a lot of code for cast Lambda to std::fuction<> and extract return type
a) static_visitor as base is  'boost'-like way
b) lamba is std

But in case if we fix visitor for lambda usage, we cannot use lambda here, because lambda with `auto` params is beginning with C++14, so code like that

```
auto l = []( const auto &v)  {}      //<---auto ?
```
It' won't work in c++11

So looks like class-style is only way for  C++11

Reply:
P.S.
it is possible to use overloading in lambda like that

```
auto overloaded_lamda = make_overloaded_lambda([] (const GMatDescr& m) {}, 
                                                                                     [] (const GOpaqDesc& o) {});
visit(overloaded_lambda, variant);
```
but it looks more ugly